python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -0,0 +1,311 @@
+                    yield next_batches
+                    next_batches = []
+            except StopIteration:
+                break
+
+        if len(next_batches) > 0:
+            yield next_batches
+
+    # Fetch the initial set of batches.
+    batch_iterator = get_next_batches()
+    try:
+        batches = next(batch_iterator)
+    except StopIteration:
+        return
+
+    block_refs = [block_ref for batch in batches for block_ref in batch.block_refs]
+    prefetcher.prefetch_blocks(block_refs)
+
+    for next_batches in batch_iterator:
+        # Prefetch the next batches.
+        block_refs = [
+            block_ref for batch in next_batches for block_ref in batch.block_refs
+        ]
+        prefetcher.prefetch_blocks(block_refs)
+
+        for batch in batches:
+            yield batch
+
+        batches = next_batches
+
+    # Yield the final set of batches.
+    for batch in batches:
+        yield batch
+
+
+def resolve_logical_batch(
+    logical_batch_iter: Iterator[LogicalBatch],
+) -> Iterator[LogicalBatch]:
+    """Resolves the block references for each logical batch."""
+    for logical_batch in logical_batch_iter:
+        logical_batch.resolve()
+        yield logical_batch
+
+
+def construct_batch_from_logical_batch(
+    resolved_logical_batch_iter: Iterator[LogicalBatch],
+    ensure_copy: bool = False,
+) -> Iterator[Batch]:
+    """Given an iterator over logical batches, returns an iterator over actual
+    constructed batches.
+
+    Args:
+        resolved_logical_batch_iter: An iterator over resolved logical batches.
+        stats: Dataset stats object used to store block batching time.
+        ensure_copy: Whether batches are always copied from the underlying base
+            blocks (not zero-copy views).
+
+    Returns:
+        An iterator over batch index and batches of the given size.
+    """
+
+    for logical_batch in resolved_logical_batch_iter:
+        output = DelegatingBlockBuilder()
+        slice_indices = [[0, None] for _ in range(len(logical_batch.blocks))]
+        if logical_batch.starting_block_idx > 0:
+            slice_indices[0][0] = logical_batch.starting_block_idx
+        if logical_batch.ending_block_idx is not None:
+            slice_indices[-1][1] = logical_batch.ending_block_idx
+
+        for i, block in enumerate(logical_batch.blocks):
+            accessor = BlockAccessor.for_block(block)
+            slice_index = slice_indices[i]
+            output.add_block(
+                accessor.slice(
+                    slice_index[0],
+                    slice_index[1]
+                    if slice_index[1] is not None
+                    else accessor.num_rows(),
+                    copy=False,
+                )
+            )
+
+        batch = output.build()
+        assert len(batch) == logical_batch.num_rows, (
+            len(batch),
+            logical_batch.num_rows,
+        )
+        if ensure_copy:
```

Comment:
We should only explicitly create a copy here if building the batch with the block builder wouldn't have already created a copy, as done here: https://github.com/ray-project/ray/blob/dee4e2eeb41441a152bb75d2fbc5faae16d6f4d7/python/ray/data/_internal/batcher.py#L143-L148

Reply:
Switched back to using `Batcher` and `ShufflingBatcher`